File: src/NubeSync.Client/Data/ChangeTracker.cs
using System;
using System.Collections.Generic;
using System.Threading.Tasks;

namespace NubeSync.Client.Data
{
    public class ChangeTracker : IChangeTracker
    {
        public Task<List<NubeOperation>> TrackAddAsync<T>(T item) where T : NubeTable
        {
            if (string.IsNullOrEmpty(item.Id))
            {
                throw new ArgumentNullException("item id");
            }

            var tableName = item.GetType().Name;
            var operations = new List<NubeOperation>
            {
                new NubeOperation()
                {
                    TableName = tableName,
                    ItemId = item.Id,
                    Type = OperationType.Added
                }
            };

            foreach (var property in item.GetProperties())
            {
                if (property.Value != default)
                {
                    operations.Add(new NubeOperation()
                    {
                        TableName = tableName,
                        ItemId = item.Id,
                        Type = OperationType.Modified,
                        Property = property.Key,
                        Value = property.Value
                    });
                }
            }

            return Task.FromResult(operations);
        }

        public Task<NubeOperation> TrackDeleteAsync<T>(T item) where T : NubeTable
        {
            if (string.IsNullOrEmpty(item.Id))
            {
                throw new ArgumentNullException("item id");
            }

            var operation = new NubeOperation()
            {
                TableName = item.GetType().Name,
                ItemId = item.Id,
                Type = OperationType.Deleted
            };

            return Task.FromResult(operation);
        }

        public Task<List<NubeOperation>> TrackModifyAsync<T>(T oldItem, T newItem) where T : NubeTable
        {
            if (string.IsNullOrEmpty(newItem.Id))
            {
                throw new ArgumentNullException("newItem id");
            }

            if (oldItem.GetType() != newItem.GetType())
            {
                throw new InvalidOperationException("Cannot compare objects of different types");
            }

            if (oldItem.Id != newItem.Id)
            {
                throw new InvalidOperationException("Cannot compare different records");
            }

            var oldProperties = oldItem.GetProperties();
            var newProperties = newItem.GetProperties();
            var tableName = newItem.GetType().Name;
            var operations = new List<NubeOperation>();

            foreach (var property in newProperties)
            {
                var oldPropertyValue = oldProperties[property.Key];
                if (oldPropertyValue != property.Value)
                {
                    operations.Add(new NubeOperation()
                    {
                        TableName = tableName,
                        ItemId = newItem.Id,
                        Type = OperationType.Modified,
                        Property = property.Key,
                        Value = property.Value,
                        OldValue = oldPropertyValue
                    });
                }
            }

            return Task.FromResult(operations);
        }
    }
}

File: samples/Clients/Flutter/nubesyncflutter/lib/_nubeSync/change_tracker.dart
import 'package:nubesyncflutter/_nubeSync/nube_client.dart';
import 'package:nubesyncflutter/_nubeSync/nube_operation.dart';
import 'package:nubesyncflutter/_nubeSync/nube_table.dart';

class ChangeTracker {
  ChangeTracker._privateConstructor();
  static final ChangeTracker instance = ChangeTracker._privateConstructor();

  Future<void> trackAdd(NubeTable item) async {
    if (item.id != null && item.id!.isEmpty) {
      throw ('cannot track add for empty id');
    }

    var type = item.runtimeType.toString();
    var operations = <NubeOperation>[];
    var operation = NubeOperation(
        tableName: item.tableName,
        itemId: item.id!,
        type: OperationType.added,
        property: '',
        value: '');
    operations.add(operation);

    var properties = item.toMap();
    for (var key in properties.keys) {
      var val = _convertValue(properties[key]);

      if (key is String && val != null && val.isEmpty) {
        print('skipping $key because it is empty');
      } else {
        operations.add(NubeOperation(
            tableName: type,
            itemId: item.id!,
            type: OperationType.modified,
            property: key,
            value: _convertValue(properties[key]),
            createdAt: item.updatedAt));
      }
    }

    await NubeClient.instance.addOperations(operations);
  }

  Future<void> trackModify(NubeTable oldItem, NubeTable newItem) async {
    if (newItem.id != null && newItem.id!.isEmpty) {
      throw ('cannot track modify for newItem with empty id');
    }

    if (oldItem.runtimeType != newItem.runtimeType) {
      throw ('cannot compare objects of different type');
    }

    if (oldItem.id != newItem.id) {
      throw ('cannot compare different records');
    }

    var oldProperties = oldItem.toMap();
    var newProperties = newItem.toMap();
    var tableName = newItem.tableName;
    var operations = <NubeOperation>[];
    var obsoleteOperations = <NubeOperation>[];

    for (var key in newProperties.keys) {
      var oldPropertyValue = oldProperties[key];
      if (oldPropertyValue != newProperties[key] && key != 'id') {
        operations.add(NubeOperation(
            tableName: tableName,
            itemId: newItem.id!,
            type: OperationType.modified,
            property: key,
            value: newProperties[key],
            oldValue: oldPropertyValue));

        // cleanup obsolete operations
        obsoleteOperations.addAll((await NubeClient.instance.getOperations())
            .where((o) =>
                o.itemId == newItem.id &&
                o.property == key &&
                o.type == OperationType.modified));
      }

      if (operations.isEmpty) {
        return;
      }

      await NubeClient.instance.addOperations(operations);
      await NubeClient.instance.deleteOperations(obsoleteOperations);
    }
  }

  Future<void> trackDelete(NubeTable item) async {
    if (item.id!.isEmpty) {
      throw ('cannot track delete for item with empty id');
    }

    var operations = <NubeOperation>[];
    operations.add(NubeOperation(
        tableName: item.tableName,
        itemId: item.id!,
        value: '',
        property: '',
        type: OperationType.deleted));
    await NubeClient.instance.addOperations(operations);

    // cleanup obsolete operations
    var obsoleteOperations = (await NubeClient.instance.getOperations())
        .where((o) => o.itemId == item.id && o.type != OperationType.deleted);
    await NubeClient.instance.deleteOperations(obsoleteOperations.toList());
  }

  String? _convertValue(dynamic value) {
    if (value is DateTime) {
      return value.toIso8601String();
    }

    return value != null ? value.toString() : null;
  }
}
